scripts/train.py

```diff
@@ -0,0 +1,190 @@
+                 weight_decay=top_config["train_params"]["optimize_params"]["weight_decay"],
+                 eps=top_config["train_params"]["optimize_params"]["eps"])
+
+scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizer,
+                                                 verbose=True)
+
+# TODO(diwei): Adjust the type and parameters of the loss
+loss = nn.CrossEntropyLoss()
+
+
+def train(model, data_loader, optimizer, loss, device='cpu'):
+    """THe interface to train the model using training data
+
+    Parameters
+    ----------
+    model : nn.Module
+        The TopFormer model used to train.
+    data_loader : DataLoader
+        The data_loader to visit the data.
+    optimizer : Adam
+        Optimizer to iterate the params.
+    loss : Loss nn.CrossEntropyLoss
+        Loss loss.
+    device : str, optional
+        Device used to evaluate, by default 'cpu'
+
+    Returns
+    -------
+    float
+        Average loss.
+    """
+    model.train()
+    epoch_loss = 0
+    model.to(device)
+    for i, batch in enumerate(data_loader):
+        tgt = batch["tgt"]
+        mem = batch["mem"]
+        gt = batch["gt"]
+
+        optimizer.zero_grad()
+        output = model(tgt.permute(1, 0, 2), mem.permute(1, 0, 2))
+        loss_result = loss(torch.max(output, dim=0)[0], gt)
+        loss_result.backward()
+        optimizer.step()
+
+        epoch_loss += loss_result.item()
+
+        # TODO(diwei): use other progressbar
+        print('step :', round((i / (len(data_loader) + + sys.float_info.epsilon)) * 100, 2),
+              '% , loss :', loss_result.item())
+
+    return epoch_loss / (len(data_loader) + sys.float_info.epsilon)
+
+
+def evaluate(model, data_loader, loss, device='cpu'):
+    """The interface to evaluate the model using validation data
+
+    Parameters
+    ----------
+    model : nn.Module
+        The TopFormer model used to evaluate.
+    data_loader : DataLoader
+        The data_loader to visit the data.
+    loss : Loss nn.CrossEntropyLoss
+        Loss loss.
+    device : str, optional
+        Device used to evaluate, by default 'cpu'.
+
+    Returns
+    -------
+    float
+        Average loss.
+    """
+    model.eval()
+    epoch_loss = 0
+    model.to(device)
+    with torch.no_grad():
+        for _, batch in enumerate(data_loader):
+            mem = batch["mem"]
+            tgt = batch["tgt"]
+            gt = batch["gt"]
+            output = model(mem.permute(1, 0, 2), tgt.permute(1, 0, 2))
+
+            loss_result = loss(torch.max(output, dim=0)[0], gt)
+            epoch_loss += loss_result.item()
+
+    return epoch_loss / len(data_loader)
+
+
+def run(total_epoch: int, best_loss: float):
+    """The interface to run the train process
+
+    Parameters
+    ----------
+    total_epoch : int
+        Total epoch num.
+    best_loss : float
+        Record the best loss.
+    """
+    train_losses, test_losses = [], []
```

Comment:
`test_` -> `validation_`.

At least to me, validation and testing are different, so I suggest that we do not mix the terms.

Hence, let's stick to the term `validation` for data used for validating the model and tuning model hyperparameters during the training phase.

Reply:
agree